src/python/pants/rules/core/test.py

```diff
@@ -25,33 +23,39 @@ class Test(Goal):
 
 
 @console_rule
-def fast_test(console: Console, addresses: BuildFileAddresses) -> Test:
-  test_results = yield [Get(TestResult, Address, address.to_address()) for address in addresses]
+def fast_test(console: Console,
+              targets: HydratedTargets,
```

Comment:
So, there is a minor optimization here which _might_ not matter, but which might be worth a TODO.

The effect of moving `HydratedTargets` into the signature like this is that `fast_test` will not start running until source globs have been expanded to hydrate all targets.

The previous formulation would only parse `BUILD` files, and would then request a test result per `Address`: the effect of that was that tests might be able to start running while source glob expansion was still ongoing. To preserve that behaviour, you could request `Optional.of(TestResult)` or some other newtype, and have `coordinator_of_tests` produce that and do the filtering.

I'd also feel a little bit better about that because of the TODO/NB in `coordinator_of_tests`: it would be nice to have the filtering logic closer to that.

Reply:
Interesting! I'll take a swipe at this tomorrow.

Reply:
Actually, did it now. I like it better this way! PTAL.